File: cmp_datasubject_frontent/src/schemas/files.ts
/**
 * Upload validation.
 *
 * Two upload paths exist — an approval proof and an import manifest — and both
 * are the input this system trusts least: one arrives from a researcher's
 * laptop, the other from a third-party capture tool.
 *
 * The client checks are a courtesy. **The server checks again and is the only
 * authority**: it re-reads the declared size against the actual bytes, sniffs
 * the content type rather than believing the browser's, and rewrites the
 * suffix. What this buys is somebody finding out their 60 MB scan is too large
 * before uploading it over a hotel connection.
 *
 * Order matters, here as on the server: refuse on size before reading, refuse
 * on type before parsing. Each step is cheaper than the one it protects.
 *
 * Mirrors `cmp.validation.files`.
 */

import { z } from "zod";

/** What one upload slot accepts. Mirrors the server's `UploadRules`. */
export interface UploadRules {
  readonly field: string;
  readonly maxBytes: number;
  readonly allowedMime: readonly string[];
  /**
   * Extensions kept on the stored file. Anything else becomes `.bin` on the
   * server, so a `.php` or `.exe` cannot be written to disk under its own name
   * whatever the client called it.
   */
  readonly allowedSuffixes: readonly string[];
  /** What to call this in a message. "an approval document", not "proof". */
  readonly describedAs: string;
}

export const PROOF: UploadRules = {
  field: "proof",
  maxBytes: 25 * 1024 * 1024,
  allowedMime: ["application/pdf", "image/png", "image/jpeg"],
  allowedSuffixes: [".pdf", ".png", ".jpg", ".jpeg"],
  describedAs: "an approval document",
};

export const MANIFEST: UploadRules = {
  field: "manifest",
  maxBytes: 25 * 1024 * 1024,
  allowedMime: ["text/csv", "application/csv", "text/plain"],
  allowedSuffixes: [".csv", ".txt"],
  describedAs: "a manifest",
};

/** "25 MB", for a message somebody has to read while annoyed. */
export function describeSize(bytes: number): string {
  if (bytes >= 1024 * 1024) return `${Math.round(bytes / (1024 * 1024))} MB`;
  return `${Math.round(bytes / 1024)} KB`;
}

function suffixOf(name: string): string {
  const dot = name.lastIndexOf(".");
  return dot === -1 ? "" : name.slice(dot).toLowerCase();
}

/**
 * A file input constrained to one slot's rules.
 *
 * Checks the suffix as well as the reported MIME type, because browsers
 * disagree about CSV: Chrome sends `text/csv`, Windows Excel installs make it
 * `application/vnd.ms-excel`, and some send nothing at all. Refusing on MIME
 * alone would reject a perfectly good manifest depending on what the user has
 * installed, which is not a rule anybody can act on.
 */
export function fileSchema(rules: UploadRules) {
  return z
    .instanceof(File, { message: `Choose ${rules.describedAs}` })
    .refine((file) => file.size > 0, "That file is empty")
    .refine(
      (file) => file.size <= rules.maxBytes,
      `Files have to be under ${describeSize(rules.maxBytes)}`,
    )
    .refine(
      (file) =>
        rules.allowedMime.includes(file.type) ||
        rules.allowedSuffixes.includes(suffixOf(file.name)),
      `Accepted formats: ${rules.allowedSuffixes.join(", ")}`,
    );
}

File: backend/api/src/cmp/validation/files.py
"""Upload validation.

Two upload paths exist — an approval proof and an import manifest — and both are
input we trust least: one arrives from a researcher's laptop, the other from a
third-party capture tool.

Checks run in this order, and the order is the point: refuse on declared size
before reading, refuse on actual size before hashing, refuse on type before
parsing. Each step is cheaper than the one it protects.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Final

from cmp.core.errors import BadRequest, ValidationFailed


@dataclass(frozen=True, slots=True)
class UploadRules:
    """What one upload slot accepts."""

    field: str
    max_bytes: int
    allowed_mime: tuple[str, ...]
    #: Extensions kept on the stored file. Anything else becomes `.bin`, so a
    #: `.php` or `.exe` cannot be written to disk under its own name whatever
    #: the client called it.
    allowed_suffixes: tuple[str, ...]


PROOF = UploadRules(
    field="proof",
    max_bytes=25 * 1024 * 1024,
    allowed_mime=("application/pdf", "image/png", "image/jpeg"),
    allowed_suffixes=(".pdf", ".png", ".jpg", ".jpeg"),
)

MANIFEST = UploadRules(
    field="manifest",
    max_bytes=25 * 1024 * 1024,
    allowed_mime=("text/csv", "application/csv", "text/plain"),
    allowed_suffixes=(".csv", ".txt"),
)

#: Evidence attached to a rights request: a holder's confirmation of what was
#: removed, or a nominee's evidence of death or incapacity. Documents and
#: images, because that is what a lab or a hospital produces.
EVIDENCE = UploadRules(
    field="evidence",
    max_bytes=25 * 1024 * 1024,
    allowed_mime=(
        "application/pdf",
        "image/png",
        "image/jpeg",
        "text/csv",
        "text/plain",
    ),
    allowed_suffixes=(".pdf", ".png", ".jpg", ".jpeg", ".csv", ".txt"),
)


def check_upload(payload: bytes, content_type: str | None, rules: UploadRules) -> None:
    """Refuse an upload that breaks the rules, naming the field that failed.

    Raises rather than returning a verdict: there is no caller that wants to
    continue with a file it has been told is unacceptable, and a boolean return
    is one forgotten `if` away from writing it anyway.
    """
    if not payload:
        raise ValidationFailed(f"The {rules.field} is empty", field=rules.field)

    if len(payload) > rules.max_bytes:
        raise BadRequest(
            f"{rules.field.capitalize()} exceeds {rules.max_bytes // (1024 * 1024)} MB",
            code="payload_too_large",
            field=rules.field,
        )

    if content_type not in rules.allowed_mime:
        raise ValidationFailed(
            f"{rules.field.capitalize()} must be one of: {', '.join(rules.allowed_mime)}",
            field=rules.field,
        )


#: What a file begins with, and what to call it when it arrives where it should
#: not have. Only the formats people actually reach for instead of a .docx are
#: listed. The commonest by a distance is a .doc: Word calls that a Word document
#: too, so somebody uploading one has not been careless, and telling them "that
#: is not a .docx" reads as the platform being wrong rather than the file.
_SIGNATURES: Final[tuple[tuple[bytes, str], ...]] = (
    (b"%PDF", "a PDF"),
    (b"\xd0\xcf\x11\xe0\xa1\xb1\x1a\xe1", "a .doc, the format Word used before 2007"),
    (b"{\\rtf", "an RTF document"),
    (b"\x89PNG\r\n\x1a\n", "a PNG image"),
    (b"\xff\xd8\xff", "a JPEG image"),
    (b"II*\x00", "a TIFF image"),
    (b"MM\x00*", "a TIFF image"),
    (b"%!PS", "a PostScript file"),
    (b"\x1f\x8b", "a gzip archive"),
    (b"<?xml", "an XML file"),
)


def describe_format(payload: bytes) -> str | None:
    """What this file appears to be, in words, or None when it is not a guess
    worth making.

    Exists to make a refusal legible, and nothing is decided by it: whoever
    calls this has already decided to refuse. A wrong guess therefore costs a
    slightly misleading sentence, never a file accepted or rejected wrongly.
    """
    for signature, name in _SIGNATURES:
        if payload.startswith(signature):
            return name
    return None


def safe_suffix(filename: str | None, rules: UploadRules) -> str:
    """The extension we are willing to store this file under.

    The client's filename is a suggestion, not an instruction. Anything outside
    the allow-list becomes `.bin` — the file is still stored, still hashed and
    still downloadable, it simply cannot claim to be something executable.
    """
    if not filename or "." not in filename:
        return ".bin"
    suffix = "." + filename.rsplit(".", 1)[-1].lower()
    return suffix if suffix in rules.allowed_suffixes else ".bin"
